File: src/main/java/org/samoxive/safetyjim/discord/commands/Warn.java
package org.samoxive.safetyjim.discord.commands;

import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import org.jooq.DSLContext;
import org.samoxive.jooq.generated.Tables;
import org.samoxive.jooq.generated.tables.records.WarnlistRecord;
import org.samoxive.safetyjim.discord.Command;
import org.samoxive.safetyjim.discord.DiscordBot;
import org.samoxive.safetyjim.discord.DiscordUtils;
import org.samoxive.safetyjim.discord.TextUtils;

import java.awt.*;
import java.util.Date;
import java.util.Scanner;

public class Warn extends Command {
    private String[] usages = { "warn @user [reason] - warn the user with the specified reason" };

    @Override
    public String[] getUsages() {
        return usages;
    }

    @Override
    public boolean run(DiscordBot bot, GuildMessageReceivedEvent event, String args) {
        Scanner messageIterator = new Scanner(args);
        JDA shard = event.getJDA();

        Member member = event.getMember();
        User user = event.getAuthor();
        Message message = event.getMessage();
        TextChannel channel = event.getChannel();
        Guild guild = event.getGuild();

        if (!member.hasPermission(Permission.KICK_MEMBERS)) {
            DiscordUtils.failMessage(bot, message, "You don't have enough permissions to execute this command!");
            return false;
        }

        if (!messageIterator.hasNext(DiscordUtils.USER_MENTION_PATTERN)) {
            return true;
        } else {
            // advance the scanner one step to get rid of user mention
            messageIterator.next();
        }

        User warnUser = message.getMentionedUsers().get(0);
        Member warnMember = guild.getMember(warnUser);

        if (user.getId().equals(warnUser.getId())) {
            DiscordUtils.failMessage(bot, message, "You can't warn yourself, dummy!");
            return false;
        }

        String reason = TextUtils.seekScannerToEnd(messageIterator);
        reason = reason.equals("") ? "No reason specified" : reason;

        Date now = new Date();

        EmbedBuilder embed = new EmbedBuilder();
        embed.setTitle("Warned in " + guild.getName());
        embed.setColor(new Color(0x4286F4));
        embed.setDescription("You were warned in " + guild.getName());
        embed.addField("Reason:", TextUtils.truncateForEmbed(reason), false);
        embed.setFooter("Warned by " + DiscordUtils.getUserTagAndId(user), null);
        embed.setTimestamp(now.toInstant());

        try {
            DiscordUtils.sendDM(warnUser, embed.build());
        } catch (Exception e) {
            DiscordUtils.sendMessage(channel, "Could not send a warning to the specified user via private message!");
        }

        DiscordUtils.successReact(bot, message);

        DSLContext database = bot.getDatabase();

        WarnlistRecord record = database.insertInto(Tables.WARNLIST,
                Tables.WARNLIST.USERID,
                Tables.WARNLIST.MODERATORUSERID,
                Tables.WARNLIST.GUILDID,
                Tables.WARNLIST.WARNTIME,
                Tables.WARNLIST.REASON)
                .values(warnUser.getId(),
                        user.getId(),
                        guild.getId(),
                        now.getTime() / 1000,
                        reason)
                .returning(Tables.WARNLIST.ID)
                .fetchOne();

        DiscordUtils.createModLogEntry(bot, shard, message, warnMember, reason, "warn", record.getId(), null, false);
        DiscordUtils.sendMessage(channel, "Warned " + DiscordUtils.getUserTagAndId(warnUser));

        return false;
    }
}

File: src/main/java/org/samoxive/safetyjim/discord/commands/Warn.kt
package org.samoxive.safetyjim.discord.commands

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import org.samoxive.safetyjim.database.SettingsEntity
import org.samoxive.safetyjim.database.WarnEntity
import org.samoxive.safetyjim.database.WarnsTable
import org.samoxive.safetyjim.discord.*
import java.awt.Color
import java.util.*

private const val ACTION_REASON = "Warning threshold exceeded."

suspend fun warnAction(guild: Guild, channel: TextChannel?, settings: SettingsEntity, modUser: User, warnUser: User, reason: String, callDepth: Int = 0) {
    val now = Date()

    val embed = EmbedBuilder()
    embed.setTitle("Warned in ${guild.name}")
    embed.setColor(Color(0x4286F4))
    embed.setDescription("You were warned in ${guild.name}")
    embed.addField("Reason:", truncateForEmbed(reason), false)
    embed.setFooter("Warned by ${modUser.getUserTagAndId()}", null)
    embed.setTimestamp(now.toInstant())

    warnUser.trySendMessage(embed.build())
    val record = WarnsTable.insertWarn(
        WarnEntity(
            userId = warnUser.idLong,
            moderatorUserId = modUser.idLong,
            guildId = guild.idLong,
            warnTime = now.time / 1000,
            reason = reason,
            pardoned = false
        )
    )

    createModLogEntry(guild, channel, settings, modUser, warnUser, reason, ModLogAction.Warn, record.id)

    if (settings.warnThreshold != 0) {
        val warnCount = WarnsTable.fetchUserActionableWarnCount(guild, warnUser)
        if (warnCount >= settings.warnThreshold) {
            val expirationDate = settings.getWarnActionExpirationDate()
            executeModAction(settings.warnAction, guild, channel, settings, modUser, warnUser, ACTION_REASON, expirationDate, callDepth)
        }
    }
}

class Warn : Command() {
    override val usages = arrayOf("warn @user [reason] - warn the user with the specified reason")

    override suspend fun run(bot: DiscordBot, event: GuildMessageReceivedEvent, settings: SettingsEntity, args: String): Boolean {
        val messageIterator = Scanner(args)

        val member = event.member!!
        val user = event.author
        val message = event.message
        val channel = event.channel
        val guild = event.guild

        if (!member.hasPermission(Permission.KICK_MEMBERS)) {
            message.failMessage("You don't have enough permissions to execute this command!")
            return false
        }

        if (args.isEmpty()) {
            return true
        }

        val (searchResult, warnUser) = messageIterator.findUser(message)
        if (searchResult == SearchUserResult.NOT_FOUND || (warnUser == null)) {
            message.failMessage("Could not find the user to warn!")
            return false
        }

        if (searchResult == SearchUserResult.GUESSED) {
            message.askConfirmation(bot, warnUser) ?: return false
        }

        if (user == warnUser) {
            message.failMessage("You can't warn yourself, dummy!")
            return false
        }

        var reason = messageIterator.seekToEnd()
        reason = if (reason == "") "No reason specified" else reason

        warnAction(guild, channel, settings, user, warnUser, reason)
        message.successReact()
        channel.sendModActionConfirmationMessage(settings, "Warned ${warnUser.getUserTagAndId()}")

        return false
    }
}
